smithy-protocols-traits/src/main/java/software/amazon/smithy/protocols/traits/Rpcv2TraitValidator.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except
+ * in compliance with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.protocols.traits;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.ServiceIndex;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Validates models implementing the {@code Rpcv2Trait} against its constraints by:
+ *
+ * - Ensuring that every entry in {@code eventStreamHttp} also appears in the {@code http} property of
+ * a protocol trait.
+ * - Ensuring that there is at least one value for the {@code format} property.
+ * - Ensuring all the {@code format} property values are valid and supported.
+ */
+@SmithyInternalApi
+public final class Rpcv2TraitValidator extends AbstractValidator {
+
+    private static final List<String> VALID_FORMATS = ListUtils.of("cbor");
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        ServiceIndex serviceIndex = ServiceIndex.of(model);
+        return model.shapes(ServiceShape.class)
+                .flatMap(service -> validateService(service, serviceIndex).stream())
+                .collect(Collectors.toList());
+    }
+
+    private List<ValidationEvent> validateService(ServiceShape service, ServiceIndex index) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        for (Trait protocol : index.getProtocols(service).values()) {
+            if (protocol instanceof Rpcv2Trait) {
```

Comment:
Then instead of needing the `ServiceIndex` and loop here, you can assert the service has the trait:

`Rpcv2Trait rpcv2Trait = service.expectTrait(Rpcv2Trait.class);`